riscv.c

```diff
@@ -773,67 +827,550 @@ static bool op_amo(struct riscv_t *rv, uint32_t inst)
 // handler for all unimplemented opcodes
 static bool op_unimp(struct riscv_t *rv, uint32_t inst UNUSED)
 {
-    rv_except_illegal_inst(rv);
+    rv_except_illegal_inst(rv, inst);
+    return false;
+}
+
+#ifdef ENABLE_RV32C
+static bool op_caddi(struct riscv_t *rv, uint16_t inst)
+{
+    uint16_t tmp =
+        (uint16_t)(((inst & FCI_IMM_12) >> 5) | (inst & FCI_IMM_6_2)) >> 2;
+    const int32_t imm = (0x20 & tmp) ? 0xffffffc0 | tmp : tmp;
+    const uint16_t rd = c_dec_rd(inst);
+
+    // dispatch operation type
+    if (rd != 0) {
+        // C.ADDI
+        rv->X[rd] += imm;
+    } else {
+        // C.NOP
+    }
+
+    // step over instruction
+    rv->PC += rv->inst_len;
+    // enforce zero register
+    if (rd == rv_reg_zero)
+        rv->X[rv_reg_zero] = 0;
+    return true;
+}
+
+static bool op_caddi4spn(struct riscv_t *rv, uint16_t inst)
+{
+    uint16_t temp = 0;
+    temp |= (inst & 0x1800) >> 7;
+    temp |= (inst & 0x780) >> 1;
+    temp |= (inst & 0x40) >> 4;
+    temp |= (inst & 0x20) >> 2;
+
+    const uint16_t imm = temp;
+    const uint16_t rd = c_dec_rdc(inst) | 0x08;
+    rv->X[rd] = rv->X[2] + imm;
+
+    rv->PC += rv->inst_len;
+    return true;
+}
+
+static bool op_cli(struct riscv_t *rv, uint16_t inst)
+{
+    uint16_t tmp = (uint16_t)((inst & 0x1000) >> 7 | (inst & 0x7c) >> 2);
+    const int32_t imm = (tmp & 0x20) ? 0xffffffc0 | tmp : tmp;
+    const uint16_t rd = c_dec_rd(inst);
+    rv->X[rd] = imm;
+
+    rv->PC += rv->inst_len;
+    return true;
+}
+
+static bool op_clui(struct riscv_t *rv, uint16_t inst)
+{
+    const uint16_t rd = c_dec_rd(inst);
+    if (rd == 2) {
+        // C.ADDI16SP
+        uint32_t tmp = (inst & 0x1000) >> 3;
+        tmp |= (inst & 0x40) >> 2;
+        tmp |= (inst & 0x20) << 1;
+        tmp |= (inst & 0x18) << 4;
+        tmp |= (inst & 0x4) << 3;
+        const uint32_t imm = (tmp & 0x200) ? (0xfffffc00 | tmp) : tmp;
+
+        if (imm != 0)
+            rv->X[rd] += imm;
+        else { /*imm==0 is reserved */
+        }
+    } else if (rd != 0) {
+        // C.LUI
+        uint32_t tmp = (inst & 0x1000) << 5 | (inst & 0x7c) << 10;
+        const int32_t imm = (tmp & 0x20000) ? (0xfffc0000 | tmp) : tmp;
+        if (imm != 0)
+            rv->X[rd] = imm;
+        else { /*imm==0 is reserved*/
+        }
+    } else {
+        // assert(!"Should be unreachbale.");
```

Comment:
Why did you comment out `assert`?

Reply:
I think I leave a bad comment. That "assert" part is not unreachable. This part may be HINTS instruction. I leave it to "do nothing" in this part such as reserved instruction.